File: Src/java/model/src/test/java/org/hl7/cql/model/ChoiceTypeTests.java
package org.hl7.cql.model;

import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.util.Arrays;
import org.junit.jupiter.api.Test;

class ChoiceTypeTests {

    @Test
    void choiceTypeIsCompatible() {
        ChoiceType first = new ChoiceType(
                Arrays.asList(new SimpleType("Period"), new SimpleType("Interval"), new SimpleType("DateTime")));

        ChoiceType second = new ChoiceType(Arrays.asList(new SimpleType("Period"), new SimpleType("DateTime")));

        assertTrue(first.isCompatibleWith(second));
        assertTrue(second.isCompatibleWith(first));

        assertTrue(first.isSuperSetOf(second));
        assertFalse(first.isSubSetOf(second));
        assertTrue(second.isSubSetOf(first));
        assertFalse(second.isSuperSetOf(first));
    }

    @Test
    void choiceTypeIsNotCompatible() {
        ChoiceType first = new ChoiceType(
                Arrays.asList(new SimpleType("Period"), new SimpleType("Interval"), new SimpleType("DateTime")));

        ChoiceType second = new ChoiceType(Arrays.asList(new SimpleType("Integer"), new SimpleType("String")));

        assertFalse(first.isCompatibleWith(second));
        assertFalse(second.isCompatibleWith(first));
        assertFalse(first.isSubSetOf(second));
        assertFalse(first.isSuperSetOf(second));
    }
}

File: cql/src/jvmTest/kotlin/org/hl7/cql/model/ChoiceTypeTests.kt
package org.hl7.cql.model

import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test

internal class ChoiceTypeTests {
    @Test
    fun choiceTypeIsCompatible() {
        val first =
            ChoiceType(setOf(SimpleType("Period"), SimpleType("Interval"), SimpleType("DateTime")))

        val second = ChoiceType(setOf(SimpleType("Period"), SimpleType("DateTime")))

        Assertions.assertTrue(first.isCompatibleWith(second))
        Assertions.assertTrue(second.isCompatibleWith(first))

        Assertions.assertTrue(first.isSuperSetOf(second))
        Assertions.assertFalse(first.isSubSetOf(second))
        Assertions.assertTrue(second.isSubSetOf(first))
        Assertions.assertFalse(second.isSuperSetOf(first))
    }

    @Test
    fun choiceTypeIsNotCompatible() {
        val first =
            ChoiceType(setOf(SimpleType("Period"), SimpleType("Interval"), SimpleType("DateTime")))

        val second = ChoiceType(setOf(SimpleType("Integer"), SimpleType("String")))

        Assertions.assertFalse(first.isCompatibleWith(second))
        Assertions.assertFalse(second.isCompatibleWith(first))
        Assertions.assertFalse(first.isSubSetOf(second))
        Assertions.assertFalse(first.isSuperSetOf(second))
    }
}
